sqlglot/generator.py

```diff
@@ -2310,7 +2310,10 @@ def star_sql(self, expression: exp.Star) -> str:
 
     def parameter_sql(self, expression: exp.Parameter) -> str:
         this = self.sql(expression, "this")
-        return f"{self.PARAMETER_TOKEN}{this}"
+        expr = self.sql(expression, "expression")
+        expr = f":{expr}" if expr else ""
```

Comment:
Why do we do `:{expr}`? The colon part should be parsed separately right? E.g.

```
>>> import sqlglot
>>> sqlglot.parse_one("a:b", "snowflake")
JSONExtract(
  this=Column(
    this=Identifier(this=a, quoted=False)),
  expression=JSONPath(
    expressions=[
      JSONPathRoot(),
      JSONPathKey(this=b)]))
```

Reply:
The parser consumes `TokenType.COLON` and the following expression, but we could override `parse_parameter()` in snowflake.py to parse just `this` ?

Reply:
I don't think consuming the colon as part of the `$` parameter is correct. E.g. in this branch we get:

```python
>>> import sqlglot
>>> sqlglot.parse_one("SELECT $1:a.b", "snowflake")
Select(
  expressions=[
    Dot(
      this=Parameter(
        this=Literal(this=1, is_string=False),
        expression=Var(this=a)),
      expression=Identifier(this=b, quoted=False))])
```

Shouldn't this be a `JSONExtract` node instead? The `COLON` token makes sense as a placeholder I think. We need to make sure to not mess up the parsing of `:variable` in Snowflake, because that's [also valid](https://docs.snowflake.com/en/sql-reference/bind-variables).

Reply:
Right, I guess both representations are correct in Snowflake but the AST is not 100% there. That's because the base parser `_parse_parameter()` seems to be consuming a series of tokens and expressions but generates just `this` even before this PR, and although I wasn't sure why I didn't want to extend the scope even more. That being said:

1. Do we want to make sure that the logic in `_parse_parameter()` is still needed in other dialects as well ? If it's not needed, we can just simplify it to parse & generate `<TokenType.PARAMETER> <this>`

2. If the parameter logic is indeed correct and needed elsewhere, would it be best that we override it in snowflake to only parse `this`? From what I see, that will leave the colon as is which will be parsed properly into a json path

Reply:
I think only hive needs to consume the `COLON` token, this is the only test that failed after commenting it out:

```sql
SELECT ${hiveconf:some_var}
```

I think I added it a while ago without realizing this would affect Snowflake's syntax. Perhaps we can try overriding it in Hive to fix this.

Reply:
We can now get rid of this logic in the snowflake generator method